packages/node_modules/@ciscospark/plugin-memberships/src/memberships.js

```diff
@@ -24,6 +36,146 @@ import {SparkPlugin, Page} from '@ciscospark/spark-core';
+        this.trigger(SDK_EVENT.CREATED, membershipCreatedEventData);
+        break;
+
+      case API_ACTIVITY_VERB.LEAVE:
+        const membershipDeletedEventData =
+          this.getMembershipEvent(activity, SDK_EVENT.DELETED);
+
+        debug(`membership "deleted" payload: \
+          ${JSON.stringify(membershipDeletedEventData)}`);
+        this.trigger(SDK_EVENT.DELETED, membershipDeletedEventData);
+        break;
+
+      case API_ACTIVITY_VERB.ADD_MODERATOR:
+      case API_ACTIVITY_VERB.REMOVE_MODERATOR:
+        const membershipUpdatedEventData =
+          this.getMembershipEvent(activity, SDK_EVENT.UPDATED);
+
+        debug(`membership "updated" payload: \
+          ${JSON.stringify(membershipUpdatedEventData)}`);
+        this.trigger(SDK_EVENT.UPDATED, membershipUpdatedEventData);
+        break;
+
+      case API_ACTIVITY_VERB.ACKNOWLEDGE:
+        const membershipSeenEventData =
+          this.getMembershipEvent(activity, SDK_EVENT.SEEN);
+
+        debug(`membership "updated" payload: \
+          ${JSON.stringify(membershipSeenEventData)}`);
+        this.trigger(SDK_EVENT.SEEN, membershipSeenEventData);
+        break;
+
+      default:
+        break;
+    }
+  },
+
+  /**
+   * Constructs the data object for an event on the memberships resource,
+   * adhering to Hydra's Webehook data structure memberships.
+   * @param {Object} activity from mercury
+   * @param {Object} event type of "webhook" event
+   * @returns {Object} constructed event
+   */
+  getMembershipEvent(activity, event) {
+    const sdkEvent = JSON.parse(JSON.stringify(this.eventEnvelope));
+    let member = '';
+
+    sdkEvent.event = event;
+    sdkEvent.data.created = activity.published;
+    sdkEvent.actorId = constructHydraId(hydraTypes.PEOPLE, activity.actor.id);
+    sdkEvent.data.roomId =
+      constructHydraId(hydraTypes.ROOM, activity.target.id);
+    sdkEvent.data.roomType =
+      activity.target.tags.includes('ONE_ON_ONE') ? 'direct' : 'group';
+    if ((activity.object.roomProperties) && (activity.object.roomProperties.isModerator)) {
+      sdkEvent.data.isModerator = true;
+    }
+    else {
+      sdkEvent.data.isModerator = false;
+    }
+    sdkEvent.data.isMonitor = false;
+    // isRoomHidden
+
+    if (activity.verb !== API_ACTIVITY_VERB.ACKNOWLEDGE) {
+      // For most memberships events the person is the 'object" or the one whose membership it is
+      member = 'object';
+    }
+    else {
+      // For a read receipt the person is the "actor" or the one who did the reading
+      member = 'actor';
+      // And the "object" is the message that was last seen
+      sdkEvent.data.lastSeenId =
+        constructHydraId(hydraTypes.MESSAGE, activity.object.id);
+    }
+    sdkEvent.data.id = constructHydraId(hydraTypes.MEMBERSHIP,
+      `${activity[member].entryUUID}:${activity.target.id}`);
+    sdkEvent.data.personId =
+      constructHydraId(hydraTypes.PEOPLE, activity[member].entryUUID);
```

Comment:
We should verify that `activity[member]` exists before continuing, otherwise we'll get crashes.

Reply:
Good point.  Code looks a little more complex now.  Will ask you for another review.